src/util/s3-downloader.ts

```diff
@@ -0,0 +1,104 @@
+import fs, {PathLike} from "fs";
+import Promise from "bluebird";
+import {promises as fsPromises} from "fs";
+import IFileDownloader from "./file-downloader";
+import {DownloadFile, DownloadFilesJob} from "../common/types";
+import * as stream from "stream";
+import {GetObjectRequest} from "aws-sdk/clients/s3";
+import {S3} from "aws-sdk";
+
+class S3Downloader implements IFileDownloader {
+    private s3Instance: S3;
+
+    constructor(s3Instance: S3) {
+        this.s3Instance = s3Instance;
+    }
+
+    assertFiles(downloadJob: DownloadFilesJob): Promise<void> {
+        const workingDir = this.assertWorkingDirectory(downloadJob.basePath, downloadJob.container);
+
+        let filePromises: Promise<string>[] = [];
+        downloadJob.files.forEach((file) => filePromises.push(this.assertFile(workingDir, file)));
+        return new Promise<void>((resolve, reject) =>
+            Promise.all(filePromises)
+                .then(() => resolve())
+                .catch((error) => reject(error))
+        );
+    }
+
+    assertWorkingDirectory(basePath: string, container: string): string {
+        const workingDir = basePath + '/' + container;
+        if (!fs.existsSync(workingDir)) {
+            fs.mkdirSync(workingDir);
+        }
+        return workingDir;
+    }
+
+    assertFile(workingDir: string, downloadFile: DownloadFile): Promise<string> {
+        const filePath = workingDir + '/' + downloadFile.fileName;
+        return new Promise<string>((resolve, reject) => {
+            S3Downloader.fileExists(filePath)
+                .then(fileExists => {
+                    if (fileExists) {
+                        resolve(filePath);
+                    } else {
+                        this.getS3Stream(downloadFile.source)
+                            .then((readStream) => S3Downloader.writeFile(readStream, filePath))
+                            .then(() => resolve(filePath))
+                            .catch(error => reject(error))
+                    }
+                })
+                .catch(error => reject(error));
+        });
+    }
+
+    getS3Stream(s3Url: string): Promise<stream.Readable> {
+        return new Promise<stream.Readable>((resolve, reject) => {
+            const s3ObjectRequest = S3Downloader.s3ObjectRequest(s3Url);
+            this.s3Instance.getObject(s3ObjectRequest)
+                .promise()
+                .then(() => resolve(this.s3Instance.getObject(s3ObjectRequest).createReadStream()))
+                .catch(error => reject(error));
```

Comment:
I don't think we need this line if we're just rejecting the error here, similar to catching a RuntimeException in java and then immediately throwing it